ENH: add simple measurements panel

Punzo opened the pull request:
@lassoan here the PR for the simple measurement panel. 

the widget is a simple version of the GUI of the markups module. The idea is to have a widget that can be easly added to a toolbar and provide a simple view of the markups/measurements with minimum controls, e.g.:

```
markupsToolbar = qt.QToolBar("markupsToolbar")
markupsToolbar.objectName = "markupsToolbar"

simpleMeasurementPanel = slicer.qSlicerSimpleMeasurementsWidget()
simpleMeasurementPanel.setMRMLScene(slicer.mrmlScene)
# simpleMeasurementPanel.setMRMLScene(slicer.mrmlScene, ["vtkMRMLMarkupsLineNode"]) to filter the type of available nodes 
simpleMeasurementPanel.setShowMeasurementDisplay(True) # default False
markupsToolbar.addWidget(simpleMeasurementPanel)
    
slicer.util.mainWindow().addToolBar(qt.Qt.RightToolBarArea, markupsToolbar)
markupsToolbar.show()
```

Punzo commented:
@lassoan the spelling check fails on the institution name, https://github.com/Slicer/Slicer/pull/6662/files#diff-eb103d99b40489c3a0f843559cbddac39dddbad596cc4abb6a646d2465188af9R5 

is there anyway to add in-line exceptions?

jamesobutler reviewed (commented):
> We should add a new "Quick measurements" module, which would be simple and would do exactly what is needed for users who just want to have quick "throwaway" measurements. This module can be implemented without changing anything in Slicer core (e.g., add a custom attribute that specifies that the markup is a "quick measurement" and the module's logic would automatically delete the markup when there are too few control points). 

re @lassoan's above comment from https://github.com/Slicer/Slicer/pull/6493#issuecomment-1212571193, since these are "throwaway" measurements because they can so quickly be re-created, I'm observing that this PR does not remove the line node if there is only 1 point placed. Upon exiting the place interaction node and the line node (aka ruler) only has 1 point, it would seem relevant for this Simple/Quick Measurements functionality to remove that incomplete ruler. 

This qSlicerSimpleMeasurementsWidget seems fine as the widget for the actual new new Simple/Quick measurements module, but I would further simplify when creating a toolbar version of this module. A toolbar version of this module would simply have options to create Line, Angle, ROI(?), and wouldn't need a table of widgets or measurements options in the toolbar. The defaults of displaying the length in the slice view for a line node is good enough. We wouldn't need to display length again in a toolbar area.

Also a UI issue of upon expanding the "Advanced" section using your code snippet in the original post, the delete button gets cut off.
![image](https://user-images.githubusercontent.com/15837524/201742311-a54c2702-7934-481a-9cc4-621fda36e5a7.png)

Punzo commented:
@jamesobutler thanks for the feedback

> since these are "throwaway" measurements because they can so quickly be re-created, I'm observing that this PR does not remove the line node if there is only 1 point placed. Upon exiting the place interaction node and the line node (aka ruler) only has 1 point, it would seem relevant for this Simple/Quick Measurements functionality to remove that incomplete ruler.

I think I forgot to reset the place mode when deleting a node, which creates the bug. I will fix it tomorrow


> This qSlicerSimpleMeasurementsWidget seems fine as the widget for the actual new new Simple/Quick measurements module, but I would further simplify when creating a toolbar version of this module. A toolbar version of this module would simply have options to create Line, Angle, ROI(?), and wouldn't need a table of widgets or measurements options in the toolbar. The defaults of displaying the length in the slice view for a line node is good enough. We wouldn't need to display length again in a toolbar area.

do you mean to remove the Advanced menu? the idea was to give the user the option to change the "default" measurement that will be displayed on the main widget (i.e. the list). I will leave Andras reply on this.


> Also a UI issue of upon expanding the "Advanced" section using your code snippet in the original post, the delete button gets cut off.

I could not reproduce this, I will tray again tomorrow.

jamesobutler commented on Modules/Loadable/Markups/Widgets/Resources/UI/qSlicerSimpleMeasurementsWidget.ui:
```diff
@@ -0,0 +1,506 @@
+<ui version="4.0">
+ <class>qSlicerSimpleMeasurementsWidget</class>
+ <widget class="qSlicerWidget" name="qSlicerSimpleMeasurementsWidget">
+  <property name="geometry">
+   <rect>
+    <x>0</x>
+    <y>0</y>
+    <width>300</width>
+    <height>864</height>
+   </rect>
+  </property>
```
Some of these hardcoded dimensions may be contributing to the "Delete" button cut off in the screenshot. My screen was 1080p tall and maybe you're using as 1200p tall monitor or more. I believe removing these defined width/height geometries should resolve this. There could be others in this UI file like this farther down in the code.

Punzo replied:
thanks!  yes I could reproduce the issue by resizing Slicer, I will fix it tomorrow

jamesobutler commented:
> do you mean to remove the Advanced menu? the idea was to give the user the option to change the "default" measurement that will be displayed on the main widget (i.e. the list). I will leave it for Andras to answer on this.

I think the “Advanced” collapsible group box is good for the widget that is the module (viewed in the Slicer module panel area). For the QToolbar that would be registered with this new “Simple Measurements” module, there shouldn’t be any advanced things. The toolbar is just the simple frequently used actions. Anything that is “Advanced” doesn’t fit into that idea of what goes into the toolbar.

jamesobutler reviewed (commented):
An additional idea for the QToolBar for simple measurements:

In addition to functionality to create the certain markups nodes, in my custom app I actually have a toolbar like this simple measurements idea which includes a qmrmlnodecombobox. I use it for line nodes for simple rulers. Upon selection of the line node in the combobox, I automatically jump slice offsets to the first control point. I also have a button to delete the currently selected node in the combobox (also an option to delete all), though this delete option could also be included in the qmrmlnodecombobox as part of the delete/rename/etc options that are usually at the bottom of this type combobox.

Punzo commented:
> > since these are "throwaway" measurements because they can so quickly be re-created, I'm observing that this PR does not remove the line node if there is only 1 point placed. Upon exiting the place interaction node and the line node (aka ruler) only has 1 point, it would seem relevant for this Simple/Quick Measurements functionality to remove that incomplete ruler.
> 
> I think I forgot to reset the place mode when deleting a node, which creates the bug. I will fix it tomorrow

fixed in https://github.com/Slicer/Slicer/pull/6662/commits/007e3798bf238a500650ba0b48aeed3e57f41050


> > > A toolbar version of this module would simply have options to create Line, Angle, ROI(?),
> > 
> > 
> > Probably we need at least the curve one. We could make it configurable.
> 
> Yeah adding closed curve could work too for things like area. I wasn’t sure at first about closed curve and I’m not sure about open curve because when there are more control points it is a bit more complex and no longer a “simple” markup. There seems to be more functionality needed to really make markups with >3 control points usable. Also, something like Point List I don’t think should be an option at all here in Simple Measurements.

added configurable allowed markups in https://github.com/Slicer/Slicer/pull/6662/commits/110148d5e7de05e671d68840062a7d52116630cb

Now you can either init the widget as:

```
markupsToolbar = qt.QToolBar("markupsToolbar")
markupsToolbar.objectName = "markupsToolbar"

simpleMeasurementPanel = slicer.qSlicerSimpleMeasurementsWidget()
simpleMeasurementPanel.setMRMLScene(slicer.mrmlScene)
markupsToolbar.addWidget(simpleMeasurementPanel)
    
slicer.util.mainWindow().addToolBar(qt.Qt.RightToolBarArea, markupsToolbar)
markupsToolbar.show()
```
or 

```
markupsToolbar = qt.QToolBar("markupsToolbar")
markupsToolbar.objectName = "markupsToolbar"

simpleMeasurementPanel = slicer.qSlicerSimpleMeasurementsWidget()
allowedMarkups = ["vtkMRMLMarkupsLineNode","vtkMRMLMarkupsAngleNode","vtkMRMLMarkupsClosedCurveNode"]
simpleMeasurementPanel.setMRMLScene(slicer.mrmlScene, allowedMarkups)
markupsToolbar.addWidget(simpleMeasurementPanel)
    
slicer.util.mainWindow().addToolBar(qt.Qt.RightToolBarArea, markupsToolbar)
markupsToolbar.show()
```

> > Also a UI issue of upon expanding the "Advanced" section using your code snippet in the original post, the delete button gets cut off.
> 
> I could reproduce this when resizing, I will fix it tomorrow.

fixed in https://github.com/Slicer/Slicer/pull/6662/commits/45a00ef88be6df648387f4231fbc588a9315b456


> > do you mean to remove the Advanced menu? the idea was to give the user the option to change the "default" measurement that will be displayed on the main widget (i.e. the list). I will leave it for Andras to answer on this.
> 
> I think the “Advanced” collapsible group box is good for the widget that is the module (viewed in the Slicer module panel area). For the QToolbar that would be registered with this new “Simple Measurements” module, there shouldn’t be any advanced things. The toolbar is just the simple frequently used actions. Anything that is “Advanced” doesn’t fit into that idea of what goes into the toolbar.

I see and I agree that as simple as possible is the best. However, I will leave it for Andras to answer on this, since he explicitly requested the feature.


> An additional idea for the QToolBar for simple measurements:
> 
> In addition to functionality to create the certain markups nodes, in my custom app I actually have a toolbar like this simple measurements idea which includes a qmrmlnodecombobox. I use it for line nodes for simple rulers. Upon selection of the line node in the combobox, I automatically jump slice offsets to the first control point. I also have a button to delete the currently selected node in the combobox (also an option to delete all), though this delete option could also be included in the qmrmlnodecombobox as part of the delete/rename/etc options that are usually at the bottom of this type combobox.

For the same reason of the previous point, I would like to keep the widget as simple as possible. Having a combobox for a ruler could be very confusing to users. The ruler should be just another measurement (i.e. a custom MarkupsLineNode with rendered hash marks).

@jamesobutler @lassoan, please let me know if any other action is needed, thanks!

jamesobutler reviewed (commented):
> We should add a new "Quick measurements" module, which would be simple and would do exactly what is needed for users who just want to have quick "throwaway" measurements. This module can be implemented without changing anything in Slicer core (e.g., add a custom attribute that specifies that the markup is a "quick measurement" and the module's logic would automatically delete the markup when there are too few control points). 

re @lassoan's above comment from https://github.com/Slicer/Slicer/pull/6493#issuecomment-1212571193, since these are "throwaway" measurements because they can so quickly be re-created, I'm observing that this PR does not remove the line node if there is only 1 point placed. If I create a line node, place a point, then end the place interaction, the line node still exists rather than the markup being deleted automatically. If you exit place interaction and the markup has too few defined control points to generate a measurement, it should automatically be cleaned up (e.g. 0/1 points for a line node, 0/1/2 points for an angle node). This would be the special functionality of markups created through this Simple/Quick measurements module.

Since this is specifically a "Measurements" module, the "line" node should also be known as "ruler". It is using the base form of a line for a ruler measurement application. In Slicer 4.X, the annotations module specifically had a "ruler" type while the new markups module as a base for various things only has an object of name "line".

Punzo commented:
> since these are "throwaway" measurements because they can so quickly be re-created, I'm observing that this PR does not remove the line node if there is only 1 point placed. If I create a line node, place a point, then end the place interaction, the line node still exists rather than the markup being deleted automatically. If you exit place interaction and the markup has too few defined control points to generate a measurement, it should automatically be cleaned up (e.g. 0/1 points for a line node, 0/1/2 points for an angle node). This would be the special functionality of markups created through this Simple/Quick measurements module.

I agree, I will add this asap


> Since this is specifically a "Measurements" module, the "line" node should also be known as "ruler". It is using the base form of a line for a ruler measurement application. In Slicer 4.X, the annotations module specifically had a "ruler" type while the new markups module as a base for various things only has an object of name "line".

I think this is out of scope this PR. Slicer customizations can always add a markup class and register it. 
On the othher hand, we could implement a child class called ruler in Slicer core (for which we will be great to have rendered hash marks in the rendering). If @lassoan give me the OK, I can implement a ruler markup class in another PR.

Punzo commented:
> > since these are "throwaway" measurements because they can so quickly be re-created, I'm observing that this PR does not remove the line node if there is only 1 point placed. If I create a line node, place a point, then end the place interaction, the line node still exists rather than the markup being deleted automatically. If you exit place interaction and the markup has too few defined control points to generate a measurement, it should automatically be cleaned up (e.g. 0/1 points for a line node, 0/1/2 points for an angle node). This would be the special functionality of markups created through this Simple/Quick measurements module.
> 
> I agree, I will add this asap
> 


done in https://github.com/Slicer/Slicer/pull/6662/commits/ea79298ca26f1f43d0cf844cb997a53884e981fa

jamesobutler reviewed (commented):
In the video below I first detail the existing behavior of the markups module where incomplete line nodes are maintained and then use the new "Simple/Quick Measurements" version introduced in this PR where right-clicking to end the interaction early of an incomplete line is cleaned up as a throwaway item 👍🏻 .

An issue seen in the video is that upon right-clicking to end the place interaction in the slice view, I get no additional pop-up for the right click content menu when using the markups module, however when right-clicking to end the place interaction of these Simple/Quick Measurement line nodes, it is incorrectly bringing up the right-click context menu as well.

https://user-images.githubusercontent.com/15837524/202262921-8e7b7d3e-674b-4f6b-ab2c-d535f63701b8.mp4

Punzo commented:
Thanks for testing!

> 
> An issue seen in the video is that upon right-clicking to end the place interaction in the slice view, I get no additional pop-up for the right click content menu when using the markups module, however when right-clicking to end the place interaction of these Simple/Quick Measurement line nodes, it is incorrectly bringing up the right-click context menu as well.
> 

I agree, this is annoying. I will fix the bug tomorrow. 

I also talked with Andras and for the advacend menu we can:
- [ ] Advanced group should be called Show Properties
- [ ] we should add a visibility option for the Show Properties group, default to False.
- [ ] Move the Delete button under the tree widget
- [ ] Delete should ask to the user if it is sure to proceed.

I will have a commit for tomorrow

Punzo commented:
> Thanks for testing!
> 
> > An issue seen in the video is that upon right-clicking to end the place interaction in the slice view, I get no additional pop-up for the right click content menu when using the markups module, however when right-clicking to end the place interaction of these Simple/Quick Measurement line nodes, it is incorrectly bringing up the right-click context menu as well.
> 
> I agree, this is annoying. I will fix the bug tomorrow.
> 
> I also talked with Andras and for the advacend menu we can:
> 
> * [x]  Advanced group should be called Show Properties
> * [x]  we should add a visibility option for the Show Properties group, default to False.
> * [x]  Move the Delete button under the tree widget
> * [x]  Delete should ask to the user if it is sure to proceed.
> 
> @jamesobutler, please let me know if you like the idea. I will have a commit for tomorrow

Done in https://github.com/Slicer/Slicer/pull/6662/commits/bb812e2e13496684d7c81bf455038a8588696deb.

However to fix additional pop-up for the right click content menu I had to modify this https://github.com/Slicer/Slicer/pull/6662/files#diff-0fb318af704d75250bc977a17277e723f1222a559dd44f262390efebfec78fa9R486 (release to press events).  

I have tested and this seems to not break anything in Slicer core (camera navigation, markups interactions and related right click content menu all works). However it could break some extensions/customizations. On the other hand, I don't see any other way to do it. @jamesobutler @lassoan.

jamesobutler commented:
I think https://github.com/Slicer/Slicer/commit/ade6a37c9035f6bd40b151f8eb1456e850da9ae0 was the previous bug fix that made right-click of markups module placement in a slice view end the interaction rather than showing the right-click context menu? Hopefully this helps in identifying the fix for this measurements stuff. It was fix by @lassoan

Punzo commented:
> I think [ade6a37](https://github.com/Slicer/Slicer/commit/ade6a37c9035f6bd40b151f8eb1456e850da9ae0) was the previous bug fix that made right-click of markups module placement in a slice view end the interaction rather than showing the right-click context menu? Hopefully this helps in identifying the fix for this measurements stuff. It was fix by @lassoan

yes, but that defined both the context menu right click and the end markups event  at _ButtonReleaseEvent_, with a smart way to processed the markups event before the camera/views ones.  The issue here is generated when I remove the markup (if has no measurements) at the end placement event. Since the markup widget does not exist anymore, the right click event will be propagated to the camera/view as well.

------------------------------------------------------------------------------

Before commit https://github.com/Slicer/Slicer/pull/6662/commits/bb812e2e13496684d7c81bf455038a8588696deb was:

Right release click with "normal" markups:
-> markups end event
-> camera/view event (blocked if there is a markup under it)

Right release click with "Quick/measurement" markups:
-> markups end event
-> delete the markup
-> camera/view event (**not** blocked because there is no markup under the mouse anymore)
-> BUG: camera/view context menu showing up

------------------------------------------------------------------------------

In https://github.com/Slicer/Slicer/pull/6662/files#diff-0fb318af704d75250bc977a17277e723f1222a559dd44f262390efebfec78fa9R486 

I changed the interaction node event to be _ButtonPressEvent_ (instead of _ButtonReleaseEvent_) in _CanProcessButtonClickEvent_; 
while it is still a _ButtonReleaseEvent_ in _ProcessButtonClickEvent_. 
This avoid the bug @lassoan.

Punzo commented:
feedback from Sara:

1) add right click to the markups list with the options:
  a) Rename
  b) Clone
  c) Delete
  d) Export to file

2) Rename Attributes menu into Measurements Settings
3) Rename Default to Display (label before combobox)
4) Put the Display combobox before the measurements list
5) Add a button to go to the markups module for more options

jamesobutler commented on Modules/Loadable/Markups/Widgets/Resources/UI/qSlicerSimpleMeasurementsWidget.ui:
```diff
@@ -0,0 +1,598 @@
+    }</string>
+       </property>
+       <property name="title">
+        <string>Add</string>
```
“Create” is used in the markups module. “Add” can imply something already exists so is not as good of a term as “Create” here.
```suggestion
        <string>Create</string>
```

Punzo replied:
I have applied the correction. Thanks for the feedback!

Punzo replied:
I have applied the correction. Thanks for the feedback!

Punzo commented:
@lassoan I got also positive feedback from @sjh26. I think this is ready for review.

Punzo commented:
commit https://github.com/Slicer/Slicer/pull/6662/commits/77b89f1208c31431cf8ec56b127c8a7cd5adfb4f 

adds a button in the markups module for toggling the simple measurement panel:

![image](https://user-images.githubusercontent.com/7985338/217535259-0a65f285-a0cc-4be9-9b0b-d81a30f52ce5.png)

Punzo commented:
current UI state as for https://github.com/Slicer/Slicer/pull/6662/commits/9c71c641273597378b96dff6822cb687d2e8d78e: 

![image](https://user-images.githubusercontent.com/7985338/217613501-e401ca3b-550b-490c-a6a1-fadceb015044.png)

lassoan reviewed (commented):
This will be awesome, but needs a bit more work.

It is no longer possible to see multiple measurements in the slice/3d views (in the markup widget's properties label). This may not be a huge issue, but then we would need a way to select what is shown there in the main Markups module GUI.

Fixed column width makes it impossible to see longer markup node names.

If a markup is in a folder then only the folder shows up in the simple measurement widget. It is not possible to expand the folder. If you add further levels, from level2, folders are expandable. It expands if arrow keys on the keyboard are used, so it is probably a display truncation issue.

Get rid of writing the measurements into the node description. It was not a clear solution just an easy workaround for showing measurement results in the subject hierarchy tree. Showing them in the markup measurements panel is nicer and makes this workaround unnecessary, and by removing manipulation of the description the design is a bit simpler and the GUI is a bit less redundant.

Make measurement results list in "Measurement display" section left-aligned (instead of centered).

Rename "Advanced markup settings" to "Edit markup properties".

Rename "Measurements" toggle button in markups module "Measurement panel".

Unrelated to your changes but would be nice to fix: make Measurement settings table in Markups module have nicer column widths. For example, Name could be auto-fit to content and Enabled could fill the available space.

In the GUI, refer to the new panel as "Markups measurement panel" (not "Simple measurement panel").

Make the displayed

jamesobutler commented:
Since this work is a bit related to this idea of "Quick Measurements" from https://github.com/Slicer/Slicer/wiki/Markups-rework#quick-measurements topic, it is currently unknown how to "Quickly" place a line node for a linear measurement from regular Slicer startup. Currently it looks like I would have to first go to the Markups Module, then press the "Markups measurement panel" button followed by pressing the "Create Line" button. 3 clicks is not really "Quick" anymore. 
- Would some form of toolbar action to directly interact with this measurements panel be something to consider? 
- Or even a toolbar for the measurements panel to quickly place a line node for a linear measurement? For example, in Slicer 4 there was an toolbutton with a menu with minimal options for placing objects for quick measurements.

Typically for a line node acting as a ruler, the measurement is displayed in the Slice view so you don't usually need to view an entire table of all line nodes as the Measurement panel provides.

lassoan commented:
I agree that the panel should be made more easily accessible. Probably somewhere on the toolbar. It is somewhat similar to the Python console and the Extensions manager, so it could be added to the toolbar that is called "Extensions". (The name does not make much sense, because the Python console is not about extensions.) Or, it could be added to the layouts toolbar, which currently has a single button. We could add the show/hide Module panel to this toolbar, too.

For me, a measurement list is an essential part of the "simple measurement" workflow, because in the views you only see those markups that happen to be in the current slice. If you want to view/edit/delete any markup you would need to manually find that markup again among hundreds of slices, which takes a lot of effort. If you find that even after making the measurement panel displayable by a single click would not be simple enough then probably we could further customize/simplify the markups toolbar.

Punzo commented:
@lassoan @jamesobutler thanks for reviewing! I have applied your comments and here a screenshot od the current status:

![image](https://user-images.githubusercontent.com/7985338/220423941-19f98d15-0c78-4e78-b391-469d261ef4bb.png)

More details:

> This will be awesome, but needs a bit more work.
> 

no problem and thanks for reviewing

> It is no longer possible to see multiple measurements in the slice/3d views (in the markup widget's properties label). This may not be a huge issue, but then we would need a way to select what is shown there in the main Markups module GUI.
> 

fixed, since I changed the design (as you suggested in the other comments) and the "primary" measurement is not anymore needed (i.e. I don't display anymore in the tree the Description field)

> Fixed column width makes it impossible to see longer markup node names.
> 

Fixed, now the panel is a Dock widget

> If a markup is in a folder then only the folder shows up in the simple measurement widget. It is not possible to expand the folder. If you add further levels, from level2, folders are expandable. It expands if arrow keys on the keyboard are used, so it is probably a display truncation issue.
> 

Fixed

> Get rid of writing the measurements into the node description. It was not a clear solution just an easy workaround for showing measurement results in the subject hierarchy tree. Showing them in the markup measurements panel is nicer and makes this workaround unnecessary, and by removing manipulation of the description the design is a bit simpler and the GUI is a bit less redundant.
> 

Done

> Make measurement results list in "Measurement display" section left-aligned (instead of centered).
> 

Done

> Rename "Advanced markup settings" to "Edit markup properties".
> 

Done


> Rename "Measurements" toggle button in markups module "Measurement panel".
> 

Done

> Unrelated to your changes but would be nice to fix: make Measurement settings table in Markups module have nicer column widths. For example, Name could be auto-fit to content and Enabled could fill the available space.
> 

Done

> In the GUI, refer to the new panel as "Markups measurement panel" (not "Simple measurement panel").

Done

>> Since this work is a bit related to this idea of "Quick Measurements" from https://github.com/Slicer/Slicer/wiki/Markups-rework#quick-measurements topic, it is currently unknown how to "Quickly" place a line node for a linear measurement from regular Slicer startup. Currently it looks like I would have to first go to the Markups Module, then press the "Markups measurement panel" button followed by pressing the "Create Line" button. 3 clicks is not really "Quick" anymore.
>> 
>> * Would some form of toolbar action to directly interact with this measurements panel be something to consider?
>> * Or even a toolbar for the measurements panel to quickly place a line node for a linear measurement? For example, in Slicer 4 there was an toolbutton with a menu with minimal options for placing objects for quick measurements.
>> 
>> Typically for a line node acting as a ruler, the measurement is displayed in the Slice view so you don't usually need to view an entire table of all line nodes as the Measurement panel provides.

>I agree that the panel should be made more easily accessible. Probably somewhere on the toolbar. It is somewhat similar to >the Python console and the Extensions manager, so it could be added to the toolbar that is called "Extensions". (The name >does not make much sense, because the Python console is not about extensions.) Or, it could be added to the layouts toolbar, >which currently has a single button. We could add the show/hide Module panel to this toolbar, too.

>For me, a measurement list is an essential part of the "simple measurement" workflow, because in the views you only see >those markups that happen to be in the current slice. If you want to view/edit/delete any markup you would need to manually >find that markup again among hundreds of slices, which takes a lot of effort. If you find that even after making the >measurement panel displayable by a single click would not be simple enough then probably we could further >customize/simplify the markups toolbar.

I moved the button nearby the go to Markups module in the Markups module. So now it's 1 click to activate the measuremet panel. Also the GUI will store the visibility now.

Please let me know, if any other action is needed. Thanks!

Punzo commented on Libs/MRML/Core/vtkMRMLSelectionNode.cxx:
```diff
@@ -46,6 +46,7 @@ vtkMRMLSelectionNode::vtkMRMLSelectionNode()
   this->SetSingletonTag("Singleton");
   this->ActivePlaceNodeClassName = nullptr;
   this->ActivePlaceNodePlacementValid = false;
+  this->ToggleSimpleMeasurementPanel = false;
```
no particular reason. Anyway I moved the meassurment panel button to the markups tool, so this is not needed anymore.

Punzo commented on Modules/Loadable/Markups/MRML/vtkMRMLMarkupsNode.cxx:
```diff
@@ -115,6 +115,7 @@ void vtkMRMLMarkupsNode::WriteXML(ostream& of, int nIndent)
 
   vtkMRMLWriteXMLBeginMacro(of);
   vtkMRMLWriteXMLBooleanMacro(locked, Locked);
+  vtkMRMLWriteXMLIntMacro(maximumNumberOfItemsInDescription, MaximumNumberOfItemsInDescription);
```
Not neede anymore since we changed the design (as you suggested in the other comments) and the "primary" measurement is not anymore needed (i.e. I don't display anymore in the tree the Description field)

Punzo commented on Modules/Loadable/Markups/MRML/vtkMRMLMarkupsNode.cxx:
```diff
@@ -2597,6 +2609,27 @@ void vtkMRMLMarkupsNode::AddMeasurement(vtkMRMLMeasurement* measurement)
 }
 
+//---------------------------------------------------------------------------
+void vtkMRMLMarkupsNode::SetMeasurementAsDefault(int id)
```
Not neede anymore since we changed the design (as you suggested in the other comments) and the "primary" measurement is not anymore needed (i.e. I don't display anymore in the tree the Description field)

Punzo commented on Modules/Loadable/Markups/MRML/vtkMRMLMarkupsNode.cxx:
```diff
@@ -2667,6 +2700,12 @@ void vtkMRMLMarkupsNode::ClearValueForAllMeasurements()
 }
 
+//---------------------------------------------------------------------------
+vtkCollection* vtkMRMLMarkupsNode::GetMeasurements()
```
Removed, It was not actually necessary. The current API was already enough.

jamesobutler reviewed (commented):
Although I'm for moving to new updated icons, I wonder if we should hold off on using new designs for the delete trash can icon and the settings gear icon. Including this change here will results in multiple trash can and gear icons being used across the application rather than a single trash can icon used throughout.

| Current | New |
|----------|------|
|![image](https://user-images.githubusercontent.com/15837524/220426432-9ad74176-663f-4fb5-bf23-64ce33aa3096.png)|![image](https://user-images.githubusercontent.com/15837524/220425946-b6e04fc1-7f5e-4278-9708-d2af66f394e0.png) | 
| ![image](https://user-images.githubusercontent.com/15837524/220426522-2a96e3d5-fef2-46f2-a7c4-e726b58feed1.png)|![image](https://user-images.githubusercontent.com/15837524/220425979-1c6cfc8e-d1e2-43c4-bdef-a5283d5d51ef.png)|

Punzo commented:
> Although I'm for moving to new updated icons, I wonder if we should hold off on using new designs for the delete trash can > icon and the settings gear icon. Including this change here will results in multiple trash can and gear icons being used across > the application rather than a single trash can icon used throughout.

I see your point, probably we should indeed hold the update of the delete and gear icons. Customization can always change the icons. I could also add helper methods for changing the icons (not sure). @lassoan what do you think about it?

lassoan commented:
I think it makes sense to decouple modernization of the icons from this pull request. If the GUI widgets have meaningful names then it is quite easy to get access and change the icon. We can continue the icons discussion at https://github.com/Slicer/Slicer/pull/5887

jamesobutler commented:
An issue with these black only SVG icons I could see is that they may be difficult when using Slicer in the Slicer Dark theme. I attempted using black SVG icons in https://github.com/Slicer/Slicer/pull/6517 and auto inverting their color when switching to the Dark style.

Punzo commented:
I will change the icons back to the "old" ones tomorrow

Punzo commented:
@jamesobutler @lassoan removed all the new system-like icons and I left only the one for the activation of the panel, i.e.:

![image](https://user-images.githubusercontent.com/7985338/220538625-1907761c-4cfa-4d71-8182-ea3a17403056.png)

Thanks for all the feedback, please let me know if it is ready for merging.

jamesobutler commented:
> In the GUI, refer to the new panel as "Markups measurement panel" (not "Simple measurement panel")

@lassoan Do you think the classes should reflect the same new name? So instead of qSlicerSimpleMeasurementsWidget it would be a qSlicerMarkupsMeasurementsWidget? Probably would be quick to find replace all on variables and file names using "SimpleMeasurements" and replace with "MarkupsMeasurements".

![image](https://user-images.githubusercontent.com/15837524/220639847-5ea274c3-f4f3-47bb-9c11-4d62b468db5b.png)
^ It is a little weird that the Markups ToolBar has QPushButton objects instead of QToolButtons tied to QActions. It obvious now that the newly added `SimpleMeasurementPanelDockWidgetViewAction` looks different from the "Open the Markups module" button to the left of it. Both result in a new markups related thing being shown, but they both have different styling.

Punzo commented:
> Probably would be quick to find replace all on variables and file names using "SimpleMeasurements" and replace with "MarkupsMeasurements".

no problem for me, if you both prefer "MarkupsMeasurements" naming, I can replace it

> It is a little weird that the Markups ToolBar has QPushButton objects instead of QToolButtons tied to QActions. It obvious now that the newly added `SimpleMeasurementPanelDockWidgetViewAction` looks different from the "Open the Markups module" button to the left of it. Both result in a new markups related thing being shown, but they have different styling.

I changed the QPushButton into an action in https://github.com/Slicer/Slicer/pull/6662/commits/8726b6d16d4b8f7c3be9785f6dbae1b29d50a0bd

![image](https://user-images.githubusercontent.com/7985338/220648058-d9dcd83b-31b1-4ecb-9783-6557931cc63e.png)

Punzo commented:
> > Probably would be quick to find replace all on variables and file names using "SimpleMeasurements" and replace with "MarkupsMeasurements".
> 
> no problem for me, if you both prefer "MarkupsMeasurements" naming, I can replace it
> 
> >

@jamesobutler  done in  https://github.com/Slicer/Slicer/pull/6662/commits/0c12ff7056dcf306e57d3268296c45455b80a00d (I also like more "MarkupsMeasurements").

Punzo commented:
@lassoan I have applied your last feedback in https://github.com/Slicer/Slicer/pull/6662/commits/d38e9ceff02f1fdff328eddbf06d1a05391f2ea4

This is ready for another review iterration if needed. Thanks!

lassoan commented:
It is almost there. A few comments:
- It is a bit confusing that all node types show up in the measurement list
- The styling of the collapsible button is unusual (we don't use yellow border elsewhere) and the line at the top is aligned to the top of the text instead of the center. In Slicer core, we should be consistent with other widgets. We can override the style sheet in custom applications.
- In measurement display section we should be able to choose what measurements are enabled/disabled.

<img width="206" alt="image" src="https://user-images.githubusercontent.com/307929/221374536-a6bb0f52-1759-4e1e-8b9f-f4751faf3c00.png">

Punzo commented:
@lassoan, thanks for retesting! I have fixed the last issues in https://github.com/Slicer/Slicer/pull/6662/commits/c232eae964b24400069ce5a8c4d7abaaa890a76d:

> It is almost there. A few comments:
> 
> * It is a bit confusing that all node types show up in the measurement list

I had forgotten to set the node types and tested the PR only with markups nodes in the scene. Nice catch, thanks!

> * The styling of the collapsible button is unusual (we don't use yellow border elsewhere) and the line at the top is aligned to the top of the text instead of the center. In Slicer core, we should be consistent with other widgets. We can override the style sheet in custom applications.

ok I have removed the groupbox style

> * In measurement display section we should be able to choose what measurements are enabled/disabled.
> 

Added the table like in the markups main module GUI.

jamesobutler commented:
@lassoan What do you think about the current state of this PR?

lassoan commented:
This is useful development that should be merged but there are a few non-trivial design changes that needs more work. I'll try to give more specific feedback this week.

jcfr commented:
@lassoan I found the branch https://github.com/Slicer/Slicer/compare/integ that seems to further improve this pull request. Especially commit https://github.com/Slicer/Slicer/commit/989add6b3c497d7a85cf78bc9075c7fa02d97a93. What do you think of rebasing this topic and including the fixup one ?

lassoan commented:
There are a couple of design changes that needs a bit more consideration.

Punzo commented:
> There are a couple of design changes that needs a bit more consideration.

If you would like to discuss them, please let me know. It would be nice to merge this (:

jcfr commented:
In the meantime, I am wondering if the the changes from commit https://github.com/Slicer/Slicer/commit/989add6b3c497d7a85cf78bc9075c7fa02d97a93 (`BUG: Fixups`) should be integrated ?

These seemed to be part a review from @lassoan and may have been "lost" in the process ?

Punzo commented:
> In the meantime, I am wondering if the the changes from commit https://github.com/Slicer/Slicer/commit/989add6b3c497d7a85cf78bc9075c7fa02d97a93 (`BUG: Fixups`) should be integrated ?
> 
> These seemed to be part of a prior review from @lassoan and may have been "lost" in the process ?

Uhm I agree, but the commit seems to be a working in progress with some To Do. Not sure if Andras is still working on it. Andreas please let me know if I should integrate it as it is, or I should wait.

lassoan commented:
I'm still working on it. There are a few higher-priority tasks on my list, but I try to at least separate and merge the straightforward changes and describe some of the issues with the other parts.

Punzo commented:
> I'm still working on it. There are a few higher-priority tasks on my list, but I try to at least separate and merge the straightforward changes and describe some of the issues with the other parts.

ok, no problem, thanks for reviewing!

Punzo committed:
ENH: Add Markups Measurement Panel

jamesobutler commented:
@lassoan Do you remember what remaining things were needed here in this PR per your last comment at https://github.com/Slicer/Slicer/pull/6662#issuecomment-1689036718? I was doing some measurement work and remembered this PR and thought it had been integrated, but arrived back here to see that it was still an open PR.

jamesobutler commented on SuperBuild/External_CTK.cmake:
```diff
@@ -81,7 +81,7 @@ if(NOT DEFINED CTK_DIR AND NOT Slicer_USE_SYSTEM_${proj})
     ${${proj}_EP_ARGS}
     GIT_REPOSITORY "${Slicer_${proj}_GIT_REPOSITORY}"
-    GIT_TAG "${Slicer_${proj}_GIT_TAG}"
+    GIT_TAG visualDICOMBrowserENH
```
This branch is still necessary for this PR? Or no since Visual DICOM browser stuff has since been integrated?

Punzo replied:
no it is not needed. I mean it was actually never needed for this branch. I just fixed it right now
